### PR TITLE
Update windows build machines

### PR DESCRIPTION
**Pull request by agocke**
Two changes: removes vs2019 in favor of vs2022

Removes 19H1 in favor of 21H1, as 19H1 is being expired.

### CODE REVIEW

File: eng/common/templates/job/execute-sdl.yml [hunk 1]
```diff
@@ -54,7 +54,7 @@ jobs:
     # If it's not devdiv, it's dnceng
     ${{ if ne(variables['System.TeamProject'], 'DevDiv') }}:
       name: NetCore1ESPool-Internal
-      demands: ImageOverride -equals Build.Server.Amd64.VS2019
```
Comment (hoyosjs) [inline eng/common/templates/job/execute-sdl.yml:57]: These changes are in eng/common.

Comment (agocke) [reply eng/common/templates/job/execute-sdl.yml:57]: Woops

**Comment by AndyAyersMS** [comment]
Seems like updating to VS2022 wasn't sufficient? The build error is still there in the `Build Cross OS Linux DAC for Windows` task
```
** Visual Studio 2022 Developer Command Prompt v17.0.6
...
[1/441] Building C object pal\src\libunwind\src\CMakeFiles\libunwind_xdac.dir\aarch64\Gtrace.c.obj
FAILED: pal/src/libunwind/src/CMakeFiles/libunwind_xdac.dir/aarch64/Gtrace.c.obj 
C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1430~1.307\bin\Hostx86\x64\cl.exe  /nologo -DBUILDENV_CHECKED=1 -DCROSS_COMPILE -DDEBUG -DDISABLE_CONTRACTS -DHAVE_CONFIG_H=1 -DHAVE_UNW_GET_ACCESSORS -DHAVE___THREAD=0 -DHOST_64BIT -DHOST_AMD64 -DHOST_WINDOWS -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_STRING=\"\" -DTARGET_64BIT -DTARGET_ARM64 -DTARGET_LINUX -DTARGET_UNIX -DUNW_REMOTE_ONLY -DURTBLDENV_FRIENDLY=Checked -D_CRT_SECURE_NO_WARNINGS -D_DBG -D_DEBUG -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_Thread_local="" -D__aarch64__ -D__linux__ -Ipal\src\libunwind\src -ID:\a\_work\1\s\src\coreclr\pal\src\libunwind\src -ID:\a\_work\1\s\src\native -ID:\a\_work\1\s\src\coreclr\pal\src\libunwind\include\tdep -ID:\a\_work\1\s\src\coreclr\pal\src\libunwind\include -Ipal\src\libunwind\include\tdep -Ipal\src\libunwind\include -ID:\a\_work\1\s\src\coreclr\pal\src\libunwind\include\win /DWIN32 /D_WINDOWS  /guard:cf /guard:ehcont -MTd   /O2 /nologo /W3 /WX /Oi /Oy- /Gm- /Zp8 /Gy /GS /fp:precise /FC /MP /Zm200 /Zc:strictStrings /Zc:wchar_t /Zc:inline /Zc:forScope /wd4065 /wd4100 /wd4127 /wd4189 /wd4200 /wd4201 /wd4245 /wd4291 /wd4456 /wd4457 /wd4458 /wd4733 /wd4838 /wd4960 /wd4961 /wd5105 /we4007 /we4013 /we4102 /we4551 /we4700 /we4640 /we4806 /w34092 /w34121 /w34125 /w34130 /w34132 /w34212 /w34530 /w35038 /w44177 /Zi /ZH:SHA_256 /source-charset:utf-8 /TC /permissive- -wd4068 -wd4146 -wd4244 -wd4267 -wd4334 -wd4311 -wd4475 -wd4477 /showIncludes /Fopal\src\libunwind\src\CMakeFiles\libunwind_xdac.dir\aarch64\Gtrace.c.obj /Fdpal\src\libunwind\src\CMakeFiles\libunwind_xdac.dir\ /FS -c D:\a\_work\1\s\src\coreclr\pal\src\libunwind\src\aarch64\Gtrace.c
D:\a\_work\1\s\src\coreclr\pal\src\libunwind\include\libunwind-aarch64.h(198): error C2061: syntax error: identifier 'alignas'
D:\a\_work\1\s\src\coreclr\pal\src\libunwind\include\libunwind-aarch64.h(199): error C2059: syntax error: '}'
D:\a\_work\1\s\src\coreclr\pal\src\libunwind\include\libunwind-aarch64.h(215): error C2079: 'uc_mcontext' uses undefined struct 'unw_sigcontext'
```

**Comment by AndyAyersMS** [comment]
Problematic line is:
```
	alignas(16) uint8_t __reserved[(66 * 8)];
```
We're including the right header `<stdalign.h>` so not clear what is wrong yet. Will need to look at preprocessor output.

**Comment by agocke** [comment]
I tried to repro this failure locally and had no luck on my win10 machine with vs2022 P1. The compiler on my machine is older though, so it might just be in an even more recent preview.

**Comment by jkotas** [comment]
> We're including the right header <stdalign.h> so not clear what is wrong yet.

`stdalign.h` comes from Windows SDK (e.g. `C:\Program Files (x86)\Windows Kits\10\Include\10.0.22000.0\ucrt`). Maybe we are using a bad Windows SDK version on CI machines?

**Comment by AndyAyersMS** [comment]
My local build is fine too. 

CI log shows CMake was able to find `<stdalign.h>` but not clear which version.
```
-- Looking for include file stdalign.h
-- Looking for include file stdalign.h - found
```

**Comment by AndyAyersMS** [comment]
Also possibly relevant: #56811.

**Comment by agocke** [comment]
Looks like this is happening in main as well: https://dev.azure.com/dnceng/public/_build/results?buildId=1629951&view=results

Probably a Windows SDK update issue

**Comment by janvorli** [comment]
The stdalign.h that's included there should be one that we generate if we don't find one. I could see in a ci failure that we have not found one on the system. So we should have generated a fake one and included it. For some reason, the CI build didn't find it. My local build also doesn't find it, generates one and it works.

**Comment by janvorli** [comment]
Ah, I can see something - on my machine, the stdalign.h was not found. However, in the CI, the stdalign.h was found (contrary to what I've said before), however in both cases the "HAVE_STDALIGN_ALIGNAS - Failed". So what happens is that the stdalign.h that's new in the SDK doesn't have the alignas in it, so when we generate the stdalign.h, the one from the SDK gets included and the alignas is missing.

**Comment by AndyAyersMS** [comment]
You might take a look at #56811 because there is some cmake policy to ensure the cmake scouting and build use the same options and such.

**Comment by agocke** [comment]
@janvorli So you'd be comfortable with a fix where we add this to fakestdalign.h? It sounds like this might be failing in servicing as well, so we'd have to port that change back quite a ways.

**Comment by hoyosjs** [comment]
should the fake check be changed to a check source compiles?

**Comment by janvorli** [comment]
> @janvorli So you'd be comfortable with a fix where we add this to fakestdalign.h?

I guess I've not explained it properly. The fakestdalign.h.in contains the alignas definition, we generate the stdalign.h from it, but since the stdalign.h is now also in the SDK, the SDK one is included first. And that one doesn't have the alignas in it (I guess, it seems to be the only explanation)

**Comment by agocke** [comment]
I see, so we won't generate it at all, if one is present already?

**Comment by hoyosjs** [comment]
> I see, so we won't generate it at all, if one is present already?

We don't generate if there's an stdalign header, that doesn't guarantee the function exists.

**Comment by agocke** [comment]
Gotcha. Is there any way to not include the system one and use our own?

**Comment by hoyosjs** [comment]
My thinking is we should have a check_source_compiles instead of header exists. If it doesn't compile, then generate the header to have the def.

**Comment by janvorli** [comment]
I don't think any of these would work. Since the SDK one doesn't have the alignas defined and we need one.

**Comment by hoyosjs** [comment]
Haven't we been faking it this whole time though?

**Comment by janvorli** [comment]
So I think the way is to define the alignas macro in the CMakeLists.txt as compiler define in case it was not found.

**Comment by janvorli** [comment]
@hoyosjs we were, but the inclusion of the SDK stdalign.h basically prevents our fake definition to be included

**Comment by hoyosjs** [comment]
I also see this

https://github.com/dotnet/runtime/blob/a41c7b333822b0ca57a2f7d3047e8cf80ac61a77/src/coreclr/pal/src/libunwind/configure.cmake#L61-L69

https://cs.github.com/dotnet/runtime/blob/b9a55b4f52243325359ced26e3d4b31ccacdc381/src/coreclr/pal/src/libunwind/config.h.in?q=HAVE_STDALIGN_ALIGNAS

That middle define seems wrong. Right?

**Comment by janvorli** [comment]
We would also need to remove alignas definition from the fake stdalign.h

**Comment by hoyosjs** [comment]
The config in libunwind also was defining it altogether. Should we just remove the middle if define. Just ask for HAVE_STDALIGN_ALIGNAS

**Comment by janvorli** [comment]
Which middle define are you referring to? I am confused :-)

**Comment by janvorli** [comment]
sigh, MSVC doesn't support definitions of macros with arguments on the command line. That's a bummer. Clang can do that.

**Comment by janvorli** [comment]
Ah, seems like this could work:
add_definitions(/FI "${CMAKE_CURRENT_BINARY_DIR}/include/stdalign.h")

**Comment by janvorli** [comment]
actually, can anyone check if something like alignas actually exists in the stdalign.h in the new SDK? Maybe we could just add command line switch to define alignas as the counterpart.

**Comment by hoyosjs** [comment]
I am spinning a new VM to use the same image as CI

**Comment by AndyAyersMS** [comment]
It would be nice if CI builds left more of an audit log (say a binlog). But we build with ninja and so you don't get many artifacts telling you how the build actually ran.

In principle you can extract a lot of the info from the PDBs, but I don't have up to date tools to crack open the relevant records anymore.... `cvdump` or `pdbdump`) -- with these you can see the exact set of includes and command line args for each compilation unit.

**Comment by hoyosjs** [comment]
So in that image:

```
-- Looking for include file stdalign.h
-- Looking for include file stdalign.h - found
...
-- Performing Test HAVE_STDALIGN_ALIGNAS
-- Performing Test HAVE_STDALIGN_ALIGNAS - Success
```

Something in stdalign defines that, while that's not the case in the CI... Even though I am using the same image (Build.Windows.Amd64.VS2022.Open@2022.0219.013611); The one in the windows SDK is:

```c++
//
// stdalign.h
//
//      Copyright (c) Microsoft Corporation. All rights reserved.
//
// The C Standard Library <stdalign.h> header.
//
#pragma once
#ifndef _INC_STDALIGN // include guard for 3rd party interop
#define _INC_STDALIGN

#include <corecrt.h>

#pragma warning(push)
#pragma warning(disable: _UCRT_DISABLED_WARNINGS)
_UCRT_DISABLE_CLANG_WARNINGS
_CRT_BEGIN_C_HEADER

#if _CRT_HAS_C11

#define alignas _Alignas
#define alignof _Alignof
#define __alignas_is_defined 1
#define __alignof_is_defined 1

#endif // _CRT_HAS_C11

_CRT_END_C_HEADER
_UCRT_RESTORE_CLANG_WARNINGS
#pragma warning(pop) // _UCRT_DISABLED_WARNINGS
#endif // _INC_STDALIGN

```

**Comment by AndyAyersMS** [comment]
Yeah, on CI machines cmake sees the header but then

```
-- Performing Test HAVE_STDALIGN_ALIGNAS
-- Performing Test HAVE_STDALIGN_ALIGNAS - Failed
```
Should we be passing `/std:c11` to MSVC C compiles, presumably to light up `_CRT_HAS_C11`?

**Comment by janvorli** [comment]
> Should we be passing `/std:c11` to MSVC C compiles

We already do that via this:
https://github.com/dotnet/runtime/blob/a82d92c36624e89e831e37515a9b0c95a4cfe183/eng/native/configurecompiler.cmake#L6-L9

**Comment by AndyAyersMS** [comment]
But it's not there in the args on the CI machine:
```
C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1430~1.307\bin\Hostx86\x64\cl.exe  /nologo -DBUILDENV_CHECKED=1 -DCROSS_COMPILE -DDEBUG -DDISABLE_CONTRACTS -DHAVE_CONFIG_H=1 -DHAVE_UNW_GET_ACCESSORS -DHAVE___THREAD=0 -DHOST_64BIT -DHOST_AMD64 -DHOST_WINDOWS -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_STRING=\"\" -DTARGET_64BIT -DTARGET_ARM64 -DTARGET_LINUX -DTARGET_UNIX -DUNW_REMOTE_ONLY -DURTBLDENV_FRIENDLY=Checked -D_CRT_SECURE_NO_WARNINGS -D_DBG -D_DEBUG -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_Thread_local="" -D__aarch64__ -D__linux__ -Ipal\src\libunwind\src -ID:\a\_work\1\s\src\coreclr\pal\src\libunwind\src -ID:\a\_work\1\s\src\native -ID:\a\_work\1\s\src\coreclr\pal\src\libunwind\include\tdep -ID:\a\_work\1\s\src\coreclr\pal\src\libunwind\include -Ipal\src\libunwind\include\tdep -Ipal\src\libunwind\include -ID:\a\_work\1\s\src\coreclr\pal\src\libunwind\include\win /DWIN32 /D_WINDOWS  /guard:cf /guard:ehcont -MTd   /O2 /nologo /W3 /WX /Oi /Oy- /Gm- /Zp8 /Gy /GS /fp:precise /FC /MP /Zm200 /Zc:strictStrings /Zc:wchar_t /Zc:inline /Zc:forScope /wd4065 /wd4100 /wd4127 /wd4189 /wd4200 /wd4201 /wd4245 /wd4291 /wd4456 /wd4457 /wd4458 /wd4733 /wd4838 /wd4960 /wd4961 /wd5105 /we4007 /we4013 /we4102 /we4551 /we4700 /we4640 /we4806 /w34092 /w34121 /w34125 /w34130 /w34132 /w34212 /w34530 /w35038 /w44177 /Zi /ZH:SHA_256 /source-charset:utf-8 /TC /permissive- -wd4068 -wd4146 -wd4244 -wd4267 -wd4334 -wd4311 -wd4475 -wd4477 /showIncludes /Fopal\src\libunwind\src\CMakeFiles\libunwind_xdac.dir\aarch64\Gtrace.c.obj /Fdpal\src\libunwind\src\CMakeFiles\libunwind_xdac.dir\ /FS -c D:\a\_work\1\s\src\coreclr\pal\src\libunwind\src\aarch64\Gtrace.c
```
...?

**Comment by janvorli** [comment]
May be a cmake bug of the specific cmake version?

**Comment by janvorli** [comment]
Sounds possible, I've found a PR in cmake around cmake and std:17 which also mentioned std:11 and it says in the change description that:

> Alternatively, a small note in the documentation for CMAKE_C_STANDARD that for MSVC this is only supported from CMake 3.22 on would help clarify that problem.

Maybe the same issue exists in the cmake version in the CI for the std:11. 
On my machine, I have cmake 3.22 and it has /std:11 on the command lines.

**Comment by janvorli** [comment]
@hoyosjs can you please check which version of cmake is in that image?

**Comment by jkoritzinsky** [comment]
I just checked CMake's source code, and they didn't add support for specifying the C language standard version to their MSVC toolchain support until CMake 3.18.2. In particular, this PR into CMake added support: https://gitlab.kitware.com/cmake/cmake/-/issues/21069

Our CMake images on the machines are likely out of date as our minimum version for Windows is 3.16.4, and I don't think the Core-Eng team updates CMake without us asking them to do so (since Kitware doesn't use Microsoft Update to push out updates).

**Comment by am11** [comment]
https://github.com/dotnet/runtime/commit/9fe775c4af7b0c3a3bed4d2c12448dd0d9f8e99f should fix the problem?

File: src/coreclr/pal/src/CMakeLists.txt [hunk 2]
```diff
@@ -230,6 +230,7 @@ add_library(coreclrpal
   ${PLATFORM_SOURCES}
   ${LIBUNWIND_OBJECTS}
 )
+target_compile_features(coreclrpal PRIVATE cxx_std_11)
```
Comment (am11) [inline src/coreclr/pal/src/CMakeLists.txt:233]: Should it be:
```suggestion
target_compile_features(coreclrpal PRIVATE c_std_11)
```

**Comment by agocke** [comment]
Sounds good, let's give it a shot

**Comment by hoyosjs** [comment]
Its the same version of the SDK (10.0.22000.0) and Compiler. The CMake version is indeed a little lower: cmake version 3.21.21080301-MSVC_2

**Comment by hoyosjs** [comment]
And even worse: CI is using the one dynamically obtainedL D:\a\_work\1\s\native-tools\bin\cmake.exe -> the one in global.json

**Comment by jkotas** [comment]
Delete the global.json entry for cmake and just use the one that comes with VS? No workarounds for old cmake necessary.

**Comment by agocke** [comment]
Are the ARM64 machines running VS? Notably this is only happening on ARM, so I wonder if that's why they have a different CMake version.

**Comment by hoyosjs** [comment]
Yes - the global.json one was creating the issues and I just repro'd. The native-dependencies install feature in arcade is being deprecated anyway, so maybe we should just remove native deps from there?

**Comment by agocke** [comment]
> Yes - the global.json one was creating the issues and I just repro'd. The native-dependencies install feature in arcade is being deprecated anyway, so maybe we should just remove native deps from there?

That sounds fine to me. We'll just have to confirm that the ARM machines have an up to date version ready to go.

**Comment by hoyosjs** [comment]
> Delete the global.json entry for cmake and just use the one that comes with VS? No workarounds for old cmake necessary.

Looks like Jan got to it before I did :)

> Are the ARM64 machines running VS? Notably this is only happening on ARM, so I wonder if that's why they have a different CMake version.

No, we always build on Windows x64, same hosts and all. It's more likely that there's plenty defines in the library that change the use of alignas depending on the architecture targeted.

**Comment by am11** [comment]
Ah, it was not the issue. Now `/std:c11` is showing up in that line from logs, but the compilation is still failed in the CI.

**Comment by AndyAyersMS** [comment]
Seems like progress though -- now we have too many definitions instead of too few:
```
D:\a\_work\1\s\artifacts\obj\coreclr\Linux.arm64.Checked\crossgen\pal\src\libunwind\include\config.h(26): error C2220: the following warning is treated as an error
D:\a\_work\1\s\artifacts\obj\coreclr\Linux.arm64.Checked\crossgen\pal\src\libunwind\include\config.h(26): warning C4005: 'alignas': macro redefinition
C:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\ucrt\stdalign.h(21): note: see previous definition of 'alignas'
```

**Comment by hoyosjs** [comment]
So the builds passed with that, but there's a lot of tests failing. Not sure if it's the change of queue (so a compiler version change) or the change in cmake logic. I think it's best to split the two changes to unblock other PRs

**Comment by am11** [comment]
For wasm build failure, I think it can be fixed by adding this line:

```xml
<_MonoCMakeBuildCommand Condition="'$(TargetsBrowser)' == 'true' and '$(HostOS)' == 'windows'">call &quot;$(RepositoryEngineeringDir)native\init-vs-env.cmd&quot; &amp;&amp; $(_MonoCMakeBuildCommand)</_MonoCMakeBuildCommand>
```
after this one:
https://github.com/dotnet/runtime/blob/95f7f7a026d74e0720d0dfdf2de799933b832df2/src/mono/mono.proj#L493

The effective `_MonoCMakeBuildCommand` on wasm was expecting that cmake is in PATH (global.json mechanism made sure of that which was deleted in last commit), but elsewhere in this file, we are not relying on that assumption and calling init-vs-env to ensure needed prerequisites are present.

cc @lambdageek

---

x64 failure is something to do with change in .yml files:
> .packages\microsoft.dotnet.helix.sdk\7.0.0-beta.22122.3\tools\Microsoft.DotNet.Helix.Sdk.MonoQueue.targets(46,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) Helix API does not contain an entry for Windows.10.Amd64.Server21H1.ES.Open

Should it be `Windows.Amd64.Server2022.Open`? that one seems to be available.

**Comment by agocke** [comment]
Still working through queue updates -- looks like there's no replacement for the ES queues yet, so I'll just end up reverting those changes. I think the PR to take through first is https://github.com/dotnet/runtime/pull/65901

**Comment by akoeplinger** [comment]
The Windows.10.Amd64.Server2022.ES.Open queue should be available now

**Comment by danmoseley** [comment]
Rebased so that hopefully we can merge this.

**Comment by danmoseley** [comment]
@hoyosjs any reason we shouldn't merge this (if green)?

**Comment by hoyosjs** [comment]
It will conflict with the arcade update. I'm happy to say  merge this and then we will resolve conflicts.

**Comment by hoyosjs** [comment]
Actually - this one needs to be in the monthly rollouts. It drops support for vs 2019.

**Comment by danmoseley** [comment]
Right, but my interpretation of comments on the other one is that VS2022 / Dev17.4 is required to change the CL errors to warnings.

**Comment by hoyosjs** [comment]
We will probably have to fold this one into that one, but yeah

**Comment by hoyosjs** [comment]
I've folded all these changes onto that one. I'll close this one.

**Pull request closed without merging**